packages/react-components/react-tree/src/components/TreeItem/useTreeItem.ts

```diff
@@ -0,0 +1,109 @@
+ * @param props - props from this instance of TreeItem
+ * @param ref - reference to root HTMLElement of TreeItem
+ */
+export const useTreeItem_unstable = (props: TreeItemProps, ref: React.Ref<TreeItemElement>): TreeItemState => {
+  const { 'aria-owns': ariaOwns, as = 'div', onKeyDown, ...rest } = props;
+
+  const { openTrees, requestOpenChange, level } = useTreeContext_unstable();
+  const { findFirstFocusable } = useFocusFinders();
+  const { targetDocument } = useFluent_unstable();
+
+  const triggerRef = useTriggerRef();
+
+  const isBranch = typeof ariaOwns === 'string';
+  const isOpen = React.useMemo(() => isBranch && openTrees.includes(ariaOwns!), [ariaOwns, openTrees, isBranch]);
+
+  const handleClick = useEventCallback((event: React.MouseEvent<TreeItemElementIntersection>) => {
+    if (isBranch) {
+      requestOpenChange({ event, open: !isOpen, type: 'click', id: ariaOwns! });
+    }
+  });
+  const handleArrowRight = (event: React.KeyboardEvent<TreeItemElementIntersection>) => {
+    if (isOpen && isBranch) {
+      // find first focusable on the subtree and focus on it
+      const subtree = targetDocument?.getElementById(ariaOwns!);
+      if (subtree) {
+        subtree && findFirstFocusable(subtree)?.focus();
+      }
+    }
+    if (isBranch && !isOpen) {
+      requestOpenChange({ event, open: true, type: 'arrowRight', id: ariaOwns! });
+    }
+  };
+  const handleArrowLeft = (event: React.KeyboardEvent<TreeItemElementIntersection>) => {
+    if (!isBranch || !isOpen) {
+      triggerRef.current?.focus();
+    }
+    if (isBranch && isOpen) {
+      requestOpenChange({ event, open: false, type: 'arrowLeft', id: ariaOwns! });
+    }
+  };
+  const handleKeyDown = useEventCallback((event: React.KeyboardEvent<TreeItemElementIntersection>) => {
+    onKeyDown?.(event);
+    if (event.isDefaultPrevented()) {
+      return;
+    }
+    switch (event.code) {
+      case ArrowRight: {
+        return handleArrowRight(event);
+      }
+      case ArrowLeft: {
+        return handleArrowLeft(event);
+      }
+    }
+  });
+  return {
+    components: {
+      root: 'div',
+    },
+    isOpen,
+    isLeaf: !isBranch,
+    root: getNativeElementProps(
+      as,
+      useARIAButtonProps(as, {
+        ...rest,
+        // casting here is required to convert union to intersection
+        ref: ref as React.Ref<TreeItemElementIntersection>,
+        'aria-owns': ariaOwns,
+        'aria-level': level,
+        // FIXME: tabster fails to navigate when aria-expanded is true
+        // 'aria-expanded': isBranch ? isOpen : undefined,
+        role: 'treeitem',
+        onClick: handleClick,
+        onKeyDown: handleKeyDown,
+      }),
+    ),
+  };
+};
+
+/**
+ * retrieves a reference to the treeitem that triggered an opening of a subtree
+ */
+function useTriggerRef() {
+  const { subtreeRef, treeRef } = useTreeContext_unstable();
+  const triggerRef = React.useRef<TreeItemElementIntersection | null>(null);
+
+  React.useEffect(() => {
```

Comment:
Instead of doing this once, it might make sense to find the trigger every time it's required, that would make sure any changes that happen during subsequent renders of the tree's shape won't impact the functionality